libs/utils/trace_analysis.py

```diff
@@ -921,7 +921,38 @@ def _integrate_square_wave(self, sq_wave):
         else:
             return sum(comp_sig.iloc[1::2].index - comp_sig.iloc[:-1:2].index)
 
-    @memoized
+    def _cropToXTimeRange(self, data):
+        """
+        Crop a dataframe containing FTrace events to the X time range specified
+        by the user through setXTimeRange().
+
+        In case self.x_min and self.x_max are set to values different than 0.0
+        and the trace duration, a first and a last entry are inserted in the
+        cropped series, such that:
+
+            - first element is a copy of the one immediately before time
+              self.x_min
+            - last element is a copy of the one immediately before time
+              self.x_max
+
+        :param data: dataframe to be cropped
+        :type data: :mod:`pandas.Series`
+        """
+        if not isinstance(data, pd.Series):
+            raise ValueError('cropping supported only for pandas.Series'\
+                             'objects!')
+
+        first = None
+        if self.x_min > 0.0:
+            first = pd.Series([data[data.index < self.x_min].iloc[-1]],
+                              index=[self.x_min])
+        last = None
+        if self.x_max < self.trace.time_range:
+            last = pd.Series([data[data.index <= self.x_max].iloc[-1]],
+                             index=[self.x_max])
+        data = data.loc[self.x_min:self.x_max]
+        return pd.concat([first, data, last])
```

Comment:
Instead of creating a series for the first and last element, you can calculate its indexes and use `data.loc` on them.  I think this function can be simplified to (not tested):
``` python
min_idx = data.loc[:self.x_min].index.max()
max_idx = data.loc[self.x_max:].index.min()
return data.loc[min_idx:max_idx]
```